python/singa/sonnx.py

```diff
@@ -989,1150 +1035,1051 @@ def from_onnx(args):
+                        val = tensor.from_numpy(val)
+                        val.to_device(self.dev)
+                        inputs.append(val)
+                    else:
+                        raise KeyError("Not found the input {} for operation {}".format(inp, node.name))
+            states = {}
+            if callable(getattr(op, "initialize",
+                                None)) and not op._initialized:
+                # init the operator
+                op.initialize(*inputs)
+                op._initialized = True
+                for key, name in node.weight_inputs.items():
+                    if key not in node.attr_inputs:
+                        # find the weights and not in the inputs
+                        states[name] = self.states[key]
+
+            # replace attrs by inputs
+            for key, name in node.attr_inputs.items():
+                if key in tensor_dict:
+                    ts = tensor_dict[key]
+                    if isinstance(ts, tensor.Tensor):
+                        ts = tensor.to_numpy(ts)
+                    states[name] = ts
+                elif key in self.states:
+                    states[name] = self.states[key]
+            # set states
+            if callable(getattr(op, "set_states", None)):
+                op.set_states(**states)
             else:
-                raise RuntimeError(
-                    "The op_name {} does not exist, please check. The available op_names are: {}"
-                    .format(op_name, [val for key, val in op_name.items()]))
-
-        # return all outputs if all_outputs==True
-        # else return last outputs
-        if all_outputs:
-            return ret_outputs
-        else:
-            return [ret_outputs[outp] for outp in final_outputs]
+                for key, value in states.items():
+                    setattr(op, key, value)
+            # run the node
+            outputs = _run_node(op, inputs)
+            # release the input tensor
+            for inp in node.inputs:
+                if inp in self.tensor_count:
+                    self.tensor_count[inp] -= 1
+                if self.tensor_count[inp] == 0:
+                    if inp in tensor_dict:
+                        del tensor_dict[inp]
+                    del self.tensor_count[inp]
+            # store the output
+            for (outp, val) in zip(node.outputs, outputs):
+                tensor_dict[outp] = val
+                if outp in outputs_dict:
+                    outputs_dict[outp] = self.to_output_tensor(val, outp)
+        return list(outputs_dict.values())
+
+
+class SONNXModel(model.Model):
+
+    def __init__(self, onnx_model):
+        """
+        Init a SIGNA Model
+        Args:
+            onnx_model (ModelProto): a loaded onnx model
+        """
+        super(SONNXModel, self).__init__()
+        self.sg_ir = prepare(onnx_model)
+        for node, operator in self.sg_ir.layers:
+            self.__dict__[node.name] = operator
```

Comment:
is each operator here a Layer instance?

Reply:
No, you can check the dict of ` _rename_operators`, for the normal operator, it creates the operator from the `autograd.py`, and for the layer operator, it creates it from `layer.py`.

Reply:
if operator is not a Layer instance, then the save_states and load_states cannot work, which works by saving the states of the sub-layers.

Reply:
At [Line 2035-2040](https://github.com/apache/singa/pull/703/files/44796e63f66f1010ac08ff2c7ba8e94160c490c4#diff-4839d7350844248bf25a18751ed06062R2035-R2040), the code will check the type of the instance, if it's from autograd, it will call `setattr(op, key, value)`, otherwise will call `op.set_states(**states)`.